cpp_utils/test/unittest/collection/database/SafeDatabaseTest.cpp

```diff
@@ -0,0 +1,971 @@
+    ASSERT_FALSE(db.modify(4, 4000));
+}
+
+/**
+ * Add some values to a <int, int> database and then erase them.
+ *
+ * STEPS:
+ * - add some values
+ * - get them
+ * - erase some
+ * - get values again expecting failures
+ * - try erase an already erased value
+ * - try erase a non existant value
+ */
+TEST(SafeDatabaseTest, test_erase)
+{
+    SafeDatabase<int, int> db;
+
+    // add some values
+    ASSERT_TRUE(db.add(1, 1000));
+    ASSERT_TRUE(db.add(2, 2000));
+    ASSERT_TRUE(db.add(3, 3000));
+
+    // get them
+    ASSERT_TRUE(db.is(1));
+    ASSERT_TRUE(db.is(2));
+    ASSERT_TRUE(db.is(3));
+
+    // erase them
+    ASSERT_TRUE(db.erase(1));
+    ASSERT_TRUE(db.erase(2));
+
+    // get values again expecting failures
+    ASSERT_FALSE(db.is(1));
+    ASSERT_FALSE(db.is(2));
+    ASSERT_TRUE(db.is(3));
+
+    // try erase an already erased value
+    ASSERT_FALSE(db.erase(1));
+
+    // try erase a non existant value
+    ASSERT_FALSE(db.erase(4));
+}
+
+/**
+ * Add some values to a <int, int> database and then modify them by adding again.
+ *
+ * STEPS:
+ * - add some values
+ * - add some values using add_or_modify
+ * - get values
+ * - modify some values using add_or_modify
+ * - check modifications took place
+ * - add some values using add_or_modify by copy
+ * - get values
+ * - modify some values using add_or_modify by copy
+ * - check modifications took place
+ */
+TEST(SafeDatabaseTest, add_or_modify)
+{
+    SafeDatabase<int, int> db;
+
+    // add some values
+    {
+        ASSERT_TRUE(db.add(1, 1000));
+        ASSERT_TRUE(db.add(2, 2000));
+    }
+
+    // add some values using add_or_modify
+    {
+        ASSERT_TRUE(db.add_or_modify(3, 3000));
+    }
+
+    // get values
+    {
+        ASSERT_EQ(db.at(1), 1000);
+        ASSERT_EQ(db.at(2), 2000);
+        ASSERT_EQ(db.at(3), 3000);
+    }
+
+    // modify some values using add_or_modify
+    {
+        ASSERT_FALSE(db.add_or_modify(1, 1500));
+    }
+
+    // check modifications took place
+    {
+        ASSERT_EQ(db.at(1), 1500);
+    }
+
+    // add some values using add_or_modify by copy
+    {
+        int k = 4;
+        ASSERT_TRUE(db.add_or_modify(k, 4000));
+    }
+
+    // get values
+    {
+        ASSERT_EQ(db.at(4), 4000);
+    }
+
+    // modify some values using add_or_modify by copy
+    {
+        int k = 1;
+        ASSERT_FALSE(db.add_or_modify(k, 1333));
+    }
+
+    // check modifications took place
+    {
+        ASSERT_EQ(db.at(1), 1333);
+    }
+}
+
+/**
+ * Access database from different threads at the same time
+ * Create a common routine of adding a value and getting it and execute it from multiple threads.
+ */
+TEST(SafeDatabaseTest, test_thread_safe)
+{
+    SafeDatabase<int, int> db;
+
+    auto routine = [&db](int i){
+        ASSERT_TRUE(db.add(i, i*1000));
+        ASSERT_TRUE(db.is(i));
+        ASSERT_EQ(db.at(i), i*1000);
+
+        int sum = 0;
+        for (const auto& it : db)
+        {
+            sum += it.second;
```

Comment:
Not used or checked

Reply:
This variable is not meant to be checked. Actually, it sums every value in the database, which you do not know what is in each thread, is random.
This test only checks that there are no TSAN issues.